File: C++/InterviewQuestions/ThreadPool_demo.cpp
// ThreadPool
//     ===> Workers
//     ===> BlockingQueue
//     ===> Task

// Producer ----> adding task to the queue
// Consumer ----> Picking the task from the queue and processing it..

#include <condition_variable>
#include <deque>
#include <functional>
#include <iostream>
#include <memory>
#include <mutex>
#include <thread>
#include <vector>

struct Task {
    virtual void execute() = 0;
    virtual ~Task() = default;
};

class RunnableTask : public Task {
public:
    explicit RunnableTask(std::function<void()> task) : task_(std::move(task)) {}

    void execute() override {
        task_();
    }

private:
    std::function<void()> task_;
};

class TaskQueue {
public:
    explicit TaskQueue(size_t capacity) : capacity_(capacity) {}

    void enqueue(std::shared_ptr<Task> task) {
        std::unique_lock<std::mutex> lock(mutex_);
        notFull_.wait(lock, [this] { return queue_.size() < capacity_; });
        queue_.push_back(std::move(task));
        notEmpty_.notify_all();  // producer asking consumer to wake up
    }

    std::shared_ptr<Task> dequeue() {
        std::unique_lock<std::mutex> lock(mutex_);
        notEmpty_.wait(lock, [this] { return !queue_.empty(); });
        std::shared_ptr<Task> task = queue_.front();
        queue_.pop_front();
        notFull_.notify_all();  // consumer asking producer to wake up
        if (queue_.empty()) {
            isEmpty_.notify_all();
        }
        return task;
    }

    void waitUntilEmpty() {
        std::unique_lock<std::mutex> lock(mutex_);
        isEmpty_.wait(lock, [this] { return queue_.empty(); });
    }

private:
    std::deque<std::shared_ptr<Task>> queue_;
    size_t capacity_;
    std::mutex mutex_;
    std::condition_variable notFull_;
    std::condition_variable notEmpty_;
    std::condition_variable isEmpty_;
};

class Worker {
public:
    explicit Worker(TaskQueue& taskQueue) : taskQueue_(taskQueue), running_(true) {}

    void shutdown() {
        running_ = false;
    }

    void run() {
        while (running_) {
            std::shared_ptr<Task> task = taskQueue_.dequeue();
            if (task != nullptr) {
                task->execute();
            }
        }
    }

private:
    TaskQueue& taskQueue_;
    bool running_;
};

class ThreadPool {
public:
    explicit ThreadPool(int numberOfThreads, size_t capacity = 10)
        : taskQueue_(capacity), numberOfThreads_(numberOfThreads), shutdown_(false) {
        createWorkers();
    }

    void submit(std::shared_ptr<Task> task) {
        taskQueue_.enqueue(std::move(task));
    }

    void shutdown() {
        shutdown_ = true;
        taskQueue_.waitUntilEmpty();  // let already-submitted tasks finish
        for (auto& worker : workers_) {
            worker->shutdown();
        }
        // wake up any workers blocked waiting for a task so they can exit
        for (size_t i = 0; i < workers_.size(); i++) {
            taskQueue_.enqueue(std::make_shared<RunnableTask>([] {}));
        }
        for (auto& thread : threads_) {
            thread.join();
        }
    }

    ~ThreadPool() {
        if (!shutdown_) {
            shutdown();
        }
    }

private:
    void createWorkers() {
        for (int i = 0; i < numberOfThreads_; i++) {
            auto worker = std::make_unique<Worker>(taskQueue_);
            Worker* workerPtr = worker.get();
            threads_.emplace_back([workerPtr] { workerPtr->run(); });
            workers_.push_back(std::move(worker));
        }
    }

    TaskQueue taskQueue_;
    int numberOfThreads_;
    bool shutdown_;
    std::vector<std::unique_ptr<Worker>> workers_;
    std::vector<std::thread> threads_;
};

int main() {
    ThreadPool pool(3);

    for (int i = 0; i < 5; i++) {
        pool.submit(std::make_shared<RunnableTask>([i] {
            std::cout << "Executing task " << i << std::endl;
        }));
    }

    pool.shutdown();

    return 0;
}

File: Python/InterviewQuestions/ThreadPool_demo.py

# ThreadPool
#     ===> Workers
#     ===> BlockingQueue
#     ===> Task

# Producer ----> adding task to the queue
# Consumer ----> Picking the task from the queue and processing it..

import threading
from abc import ABC, abstractmethod
from collections import deque
from typing import Callable, Deque, List, Optional


class Task(ABC):
    @abstractmethod
    def execute(self) -> None:
        ...


class RunnableTask(Task):
    def __init__(self, task: Callable[[], None]):
        self.task = task

    def execute(self) -> None:
        self.task()


class TaskQueue:
    def __init__(self, capacity: int):
        self.queue: Deque[Task] = deque()
        self.capacity = capacity
        self.lock = threading.Lock()
        self.not_full = threading.Condition(self.lock)
        self.not_empty = threading.Condition(self.lock)
        self.is_empty = threading.Condition(self.lock)

    def enqueue(self, task: Task) -> None:
        with self.not_full:
            while len(self.queue) == self.capacity:
                self.not_full.wait()
            self.queue.append(task)
            self.not_empty.notify_all()  # producer asking consumer to wake up

    def dequeue(self) -> Optional[Task]:
        with self.not_empty:
            while not self.queue:
                self.not_empty.wait()
            task = self.queue.popleft()
            self.not_full.notify_all()  # consumer asking producer to wake up
            if not self.queue:
                self.is_empty.notify_all()
            return task

    def wait_until_empty(self) -> None:
        with self.is_empty:
            while self.queue:
                self.is_empty.wait()


class Worker:
    def __init__(self, task_queue: TaskQueue):
        self.task_queue = task_queue
        self.running = True

    def shutdown(self) -> None:
        self.running = False

    def run(self) -> None:
        while self.running:
            task = self.task_queue.dequeue()
            if task is not None:
                task.execute()


class ThreadPool:
    def __init__(self, number_of_threads: int, capacity: int = 10):
        self.task_queue = TaskQueue(capacity)
        self.number_of_threads = number_of_threads
        self.workers: List[Worker] = []
        self.threads: List[threading.Thread] = []
        self.is_shutdown = False
        self._create_workers()

    def _create_workers(self) -> None:
        for _ in range(self.number_of_threads):
            worker = Worker(self.task_queue)
            thread = threading.Thread(target=worker.run, daemon=True)
            self.workers.append(worker)
            self.threads.append(thread)
            thread.start()

    def submit(self, task: Task) -> None:
        self.task_queue.enqueue(task)

    def shutdown(self) -> None:
        self.is_shutdown = True
        self.task_queue.wait_until_empty()  # let already-submitted tasks finish
        for worker in self.workers:
            worker.shutdown()
        # wake up any workers blocked waiting for a task so they can exit
        for _ in self.workers:
            self.task_queue.enqueue(RunnableTask(lambda: None))
        for thread in self.threads:
            thread.join()


def main() -> None:
    pool = ThreadPool(number_of_threads=3)

    for i in range(5):
        pool.submit(RunnableTask(lambda i=i: print(f"Executing task {i}")))

    pool.shutdown()


if __name__ == "__main__":
    main()
